src/components/RightSidebar/TableOfContents.tsx

```diff
@@ -3,15 +3,67 @@ import type { FunctionalComponent } from 'preact';
 import { useState, useEffect, useRef } from 'preact/hooks';
 import './TableOfContents.css';
 
+interface Heading {
+	depth: number;
+	slug: string;
+	text: string;
+}
 interface Props {
-	headings: { depth: number; slug: string; text: string }[];
+	headings: Heading[];
 	labels: {
 		onThisPage: string;
 		overview: string;
 	};
 	isMobile?: boolean;
 }
 
+interface TocItem extends Heading {
+	children: TocItem[];
+}
+
+function diveChildren(item: TocItem, depth: number): TocItem[] {
+	if (depth === 1) {
+		return item.children;
+	} else {
+		// e.g., 2
+		return diveChildren(item.children[item.children.length - 1], depth - 1);
+	}
+}
+function generateToc(headings: Heading[]) {
+	let toc: Array<TocItem> = [];
+
+	for (let heading of headings) {
+		if (toc.length === 0) {
+			toc.push({
+				...heading,
+				children: [],
+			});
+		} else {
+			const lastItemInToc = toc[toc.length - 1];
+			if (heading.depth < lastItemInToc.depth) {
+				throw new Error(`Orphan heading found: ${heading.text}.`);
```

Comment:
Because this Preact code will run in the browser, it would be good to avoid throwing errors that could break there. One option would be to move all this `generateToc` logic into an Astro component that wraps this one. That way we could safely throw at build time. It would also have the advantage of having all this code run only at build time and never be shipped to users. What do you think?

Reply:
Makes sense. However I have no idea how to make it with another Astro component since we can't use Astro component inside tsx. Wrapping `TableOfContents.tsx` inside Astro component would result some warnings since we pass `client:media` to an Astro component.

Anyway, I've refactor the `TableOfContents.tsx` component to receive a `toc` prop received at build time. Note that since there's no more easy way to get the `currentHeading` from `toc` data, I had to refactor `currentID` state into `currentHeading`.